Add rendering layers; other internal interface changes and cleanups

mik3y opened the pull request:
Yo Garth! I had a PTO day and a lot of travel, so made use of some disconnected time to tinker on a few things. This PR is a bit bigger than ideal, as it covers a few fronts, but the individual commits are all cleanly distinct. Aside from a few small bugfixes, here are the major changes:

1. Driver interface: Weights are always on a scale of `0.0 - 1.0`
2. Processor interface: LED color values are always on a scale of `0 - 1024`, and global brightness is handled & exposed by the controller
3. Devserver: Fixed weight injection
4. Controller: Run the show at single framerate & remove fps from the processor interface
5. Add rendering layers

You'll definitely want to look at commits & their descriptions one by one. For the FPS change, that's probably the biggest one we haven't really discussed, so if you have major feedback, no prob, I'm also happy to break out anything that needs to be.

mik3y commented:
Oh, one more note: For the RenderLayer change, you can try enabling it with a local patch along the following lines, to install an overlay processor:

```diff
--- a/floor/floor/controller/controller.py
+++ b/floor/floor/controller/controller.py
@@ -53,7 +53,7 @@ class Controller(object):
         # Ordered dict of layers to render, bottom-most layer first.
         self.layers = OrderedDict((
             ('playlist', PlaylistRenderLayer(playlist=self.playlist, all_processors=self.all_processors)),
-            ('overlay2', ProcessorRenderLayer()),
+            ('overlay2', ProcessorRenderLayer(processor=self.all_processors['Zap']())),
             ('overlay1', ProcessorRenderLayer()),
         ))
```

(We'll definitely want to update the benchmarker to support overlays!)

garthwebb commented on floor/run-show.py:
```diff
@@ -114,7 +116,20 @@ def main():
         sys.exit(1)
 
-    playlist = Playlist(args.playlist, args.processor_name)
+    if not (bool(args.playlist) ^ bool(args.processor_name)):
```
Hold on there XORo!  There is a default for args.playlist, so its always defined.  If we remove the default, we'd have to specify the default every time which is a little annoying.  I think I'd prefer having --processor override --playlist so that we can keep the default for this.  The alternative might be to throw this error if both are set AND the playlist isn't set to the default value.

mik3y replied:
Ah good call, yes, I shall banish captain XORo and make it do the right thing. 😆

garthwebb commented on floor/floor/processor/utils.py:
```diff
@@ -14,40 +14,63 @@ class clocked(object):
 
             # Too soon, return current frame.
-            if self.last_time is not None and (now - self.last_time) < frame_period:
+            if self.current_frame and self.next_time and now < self.next_time:
```
Hmm, this doesn't work for the Ripple pattern since it uses the annotation on a method that doesn't return anything (and therefore it never returns here).  I think we could probably remove `self.current_frame` here since `self.next_time` won't be defined until its run through this method once.

garthwebb commented on floor/floor/processor/utils.py:
```diff
@@ -14,40 +14,63 @@ class clocked(object):
+
+            fps_deadline = None
+            if self.frames_per_second:
+                fps_frame_period = 1.0 / self.frames_per_second
```
I think it would be good to move this out of the method so its only calculated once.

mik3y replied:
One challenge with that is it becomes effectively static / can no longer dynamically change. I don't think fps-twiddling will be _that_ important to a _normal_ run-of-show, but perhaps very useful when debugging new processors and/or new drivers. Could be micro-optimization territory too (i.e. inarguably better but to ultimately no noticeable effect); maybe we leave it for now and take a closer look in an overall benchmarking pass?

garthwebb replied:
Yah, I don't want to micro optimize too much, but I think when its trivial (unlike the non-trivial frames_per_beat comment below) we should take the opportunity.

Since frames_per_second is never directly used or changed (this can only be changed via the annotation, so static at compile time anyway right?), this would be easy to put in the constructor, e.g.:

self.fps_frame_period = 1.0 / frames_per_second

mik3y replied:
Oh! You're right - this is _decorator's_ target FPS, not the show. SGTM.

garthwebb commented on floor/floor/processor/utils.py:
```diff
@@ -14,40 +14,63 @@ class clocked(object):
+            if self.frames_per_beat:
+                bpm = context.bpm or 120.0
+                beats_per_second = 60.0/bpm
+                bps_frame_period = beats_per_second / float(self.frames_per_beat)
```
I'm not sure how, but it would be good to have this calculated outside this method so it only happens once.  I know this is keyed off `context.bpm` which can change, but at 24fps, it would be nice to not have to do 48 divisions per second that are always the same value.

garthwebb commented on floor/floor/util/color_utils.py:
```diff
@@ -121,6 +121,13 @@ def hex_to_rgb(value):
     return tuple(int(value[i:i + lv // 3], 16) for i in range(0, lv, lv // 3))
 
 
+def blend_pixel_nonalpha(bottom, top):
```
Is the idea to have a bunch of `blend_pixel_` methods?  For `nonalpha`, maybe `copy` instead to describe what's happening rather than a quality of the pixels involved?  Or maybe even `blend_mode_copy`?

I was trying to find some standard mixing names and found this:

http://illu.tistory.com/1254

It seems to assume the `float a, float b` values are [0.0, 1.0] which works out.  These would probably be a good TODO for adding later.

mik3y replied:
+1 to `copy` over blend! I didn't have other methods in mind but wanted some sort of descriptive name to say "yo this just treats black as alpha 1.0", like you suggestion better.

garthwebb commented on floor/floor/controller/controller.py:
```diff
@@ -149,55 +126,46 @@ def run_forever(self):
+            if leds and last_leds:
+                # Copy the returned buffer, since we'll likely be blending into it
+                # and cannot assume ownership.
+                leds = leds[:]
```
Is the assumption that `layer.render` might have kept a reference to the leds it returned to use in the next call to `render`?  Maybe instead there could be a `new_leds = [(0,0,0)] * 64` defined outside the for so we wouldn't have to do an array copy here that will be overwritten immediately.  Below it would then do:
```python
new_leds[idx] = blend...
```

mik3y replied:
> Is the assumption that layer.render might have kept a reference to the leds it returned to use in the next call to render?

Yes, since `render` returns a mutable value, controller must either assume/document ownership or non-ownership of it. Controller could demand ownership but that forces every processor to copy the buffer every frame, which may be a pain / inefficient for simplistic processors. (A lot of my dumb processors just cycle through some static frames, where mutation by the controller would do very unexpected things.)

> Maybe instead there could be a new_leds = [(0,0,0)] * 64 defined outside the for so we wouldn't have to do an array copy here that will be overwritten immediately.

Good idea, done in 58142c2

garthwebb commented on floor/floor/controller/controller.py:
```diff
@@ -149,55 +126,46 @@ def run_forever(self):
+            last_leds = leds
+
+        # If no layers are enabled, `leds` will be None and we shouldn't update the driver.
+        # TODO(mikey): Should we render a 'default' pattern in this case?
```
I don't think so;  it should be possible for the floor to just be "off" and not displaying anything ... unless I'm misunderstanding this.

mik3y commented:
Thanks for the review! I think I addressed all suggestions, also I rebased all the changes on account of some funky commit dates (that were making GitHub render them out-of-order).

mik3y commented:
I'm probably pushing my luck (and your patience!) here, tacked on a few _small_ but very related changes:

* 93cb887 - adds the API support for twiddling processor layers
* cb477d6 - normalizes pixel values (i.e. treat really really small values as zero)
* ded0903 - better error when you goof and make two process with same [class] name

that ought to do it, I think!

mik3y commented:
Ach. One more:

* 17969a3 - make sure devserver uses properly scaled component colors

Sorry about the size of this change... smaller PRs next time I swear.

garthwebb commented:
Quick comment; thanks for all these changes!  I'm going through them now.

As for my obsession with divisions and multiplications; I think the biggest single gain is going to be moving the LED-write/weight-read into a separate thread.  However, I'd like to knock out as many unnecessary operations in a loop when its easy/direct and avoid a death by a thousand cuts situation.  I think these layer mechanics have the possibility of quickly eating into our headroom, and it'd be nice to have the easy fixes out of the way when we profile.

mik3y commented:
> As for my obsession with divisions and multiplications; I think the biggest single gain is going to be moving the LED-write/weight-read into a separate thread. However, I'd like to knock out as many unnecessary operations in a loop when its easy/direct and avoid a death by a thousand cuts situation.

100%! - no sense having needless stuff in the critical path, think I was just confused about your first suggestion (I didn't spend enough time picking this change back up / remember what I was doing, heh).

garthwebb commented on floor/floor/driver/base.py:
```diff
@@ -15,16 +12,11 @@ def __init__(self, driver_args):
-        :return:
+        Returns the last retrieved list of weight values.
+
+        Weights are a value between 0.0 and 1.0.
```
It doesn't have to be in this PR, but after sleeping on it I feel like weight values should really be the integers 0 or 1.  Even if the sensors were consistent enough to return the same values for the same step, I can't think of anything meaningful we could do with that in a pattern with 8 people stopping around at once.

Plus the response is pretty non-linear, so most adults are going to drive the value from 0.0 to 0.9.  Getting rid of a nuanced weight value should hopefully simplify things a bit.

mik3y replied:
Even simpler! I'll make the change here, since we're introducing it in this PR.

garthwebb commented on floor/floor/server/static/playlist-controls.js:
```diff
@@ -146,6 +146,23 @@ function handleTempoUpdate(tempoData) {
+
+function handleBrightnessUpdate(brightnessData) {
+    var brightness = document.getElementById('brightness');
+    brightness.value = brightnessData.brightness;
```
Is it necessary to update the dropdown menu here since the user has just changed it to the value they want?  If yes, then we should make sure `brightnessData.brightness` returns `1.0`.  Right now its returning `1`, so when you select "100%" from the interface, it updates the server, then the response sets the dropdown to a non-existent option and blanks it out.

garthwebb replied:
I could see us having the periodic status call update brightness as well so that it matches any updates coming from another source (e.g. the midi controller)

mik3y replied:
Added this to `getStatus` - it will be continuously sync'd along with the others stuff :+1:

mik3y commented:
@garthwebb PTAL (I guess our branch rules will make you re-approve); think I got it all!

garthwebb commented on floor/floor/driver/raspberry.py:
```diff
@@ -165,7 +165,7 @@ def process_bytes(self, data_bytes):
             self.value_floor[position] = min(self.value_floor[position], value)
 
-            new_buf[position] = value
+            new_buf[position] = 1 if value else 0
```
I think this will need to change to do some noise filtering on the value, but I'll do that in another request later.

garthwebb commented on floor/floor/server/server.py:
```diff
@@ -210,7 +211,7 @@ def api_brightness():
         controller.set_brightness(brightness)
 
-    return jsonify({'brightness': controller.brightness})
+    return jsonify({'brightness': float(controller.brightness)})
```
Does this work for you?  I still seem to be getting `1` back at the browser rather than `1.0`.

mik3y replied:
Works for me, hrm..

garthwebb replied:
ok, the raw result has it as `0.0` but the preview tab of chrome dev tools simplifies it to just `0` and from the web interface, its still not showing `100%` when it should. Are you using Chrome?

garthwebb replied:
Ok, treating it like a string works for me:
```
-    return jsonify({'brightness': float(controller.brightness)})
+    return jsonify({'brightness': '{:.1f}'.format(controller.brightness)})
```

mik3y replied:
Shouldn't be necessary? Maybe you need to restart your server. ac30fdd called `toFixed` on the numeric value, which will stringify it client side..

garthwebb replied:
I've restarted it for each change and refreshed the page ... but I just did a shift-reload and that seemed to have the desired effect.
